File: php_payway/tests/PaywayTransactionServiceTest.php
<?php

use PhpPayway\Enumeration\PaywayPaymentOption;
use PhpPayway\Enumeration\PaywayTransactionType;
use PhpPayway\enumeration\PaywayTransactionCurrency;
use PhpPayway\Models\PaywayMerchant;
use PhpPayway\Models\Requests\PaywayCheckTransaction;
use PhpPayway\Models\requests\PaywayCreateTransaction;
use PhpPayway\Models\Requests\PaywayTransactionItem;
use PhpPayway\Services\PaywayClientFormRequestService;
use PhpPayway\Services\PaywayTransactionService;

uses()->group('PaywayTransactionService');

test('', function () {
    $dotenv = Dotenv\Dotenv::createImmutable(__DIR__);
    $dotenv->load();

    $merchant = new PaywayMerchant(
        merchantID: $_ENV['ABA_PAYWAY_MERCHANT_ID'] ?? '',
        merchantApiName: $_ENV['ABA_PAYWAY_MERCHANT_NAME'] ?? '',
        merchantApiKey: $_ENV['ABA_PAYWAY_API_KEY'] ?? '',
        baseApiUrl: $_ENV['ABA_PAYWAY_API_URL'] ?? '',
        refererDomain: "http://mylekha.app",
    );

    $service = new PaywayTransactionService(merchant: $merchant);

    $tran_id = $service->uniqueTranID();
    $items = [
        new PaywayTransactionItem(name: "ទំនិញ 1", quantity: 1, price: 1,),
        new PaywayTransactionItem(name: "ទំនិញ 2", quantity: 1, price: 2,),
        new PaywayTransactionItem(name: "ទំនិញ 3", quantity: 1, price: 3,),
    ];
    $transaction = new PaywayCreateTransaction(
        tran_id: $tran_id,
        req_time: $service->uniqueReqTime(),
        amount: 6.00,
        items: $items,
        firstname: 'Miss',
        lastname: 'My Lekha',
        phone: '010464144',
        email: 'support@mylekha.app',
        option: PaywayPaymentOption::abapay_deeplink,
        type: PaywayTransactionType::purchase,
        currency: PaywayTransactionCurrency::USD,
        return_url: "https://stage.mylekha.app",
        shipping: 0.0,
    );
    $form_service = new PaywayClientFormRequestService(merchant: $merchant);
    $request_options = $form_service->generateCreateTransactionFormData(transaction: $transaction);

//    dd($request_options);

    $createResponse = $service->createTransaction(transaction: $transaction);


    \PHPUnit\Framework\assertEquals($createResponse->abapay_deeplink != null, true,
        message: "the deeplink should be a string according to docs");

    $checkTransaction = $transaction = new PaywayCheckTransaction(
        tran_id: $tran_id,
        req_time: $service->uniqueReqTime());
    $checkResponse = $service->checkTransaction(transaction: $checkTransaction);

//    dd($checkResponse);
    \PHPUnit\Framework\assertEquals($checkResponse->status == 2, true,
        message: "the new transaction created should be pending or status 2");

    $isValidate = $service->isTransactionCompleted(transaction: $checkTransaction);
//    dd($isValidate);

    \PHPUnit\Framework\assertEquals($isValidate, false,
        message: "the new transaction created should be pending");

})->group("PaywayTransactionService");

File: dart_payway/test/payway_transaction_service_test.dart
import 'dart:math';

import 'package:dart_payway/dart_payway.dart';
import 'package:test/test.dart';
import 'package:dotenv/dotenv.dart';
import 'dart:io' as io;

void main() {
  group("load env and test", () {
    late PaywayTransactionService service;
    setUpAll(() {
      io.HttpOverrides.global = null;

      var env = DotEnv(includePlatformEnvironment: true)..load();

      service = PaywayTransactionService(
          merchant: PaywayMerchant(
        merchantID: env['ABA_PAYWAY_MERCHANT_ID'] ?? '',
        merchantApiName: env['ABA_PAYWAY_MERCHANT_NAME'] ?? '',
        merchantApiKey: env['ABA_PAYWAY_API_KEY'] ?? '',
        baseApiUrl: env['ABA_PAYWAY_API_URL'] ?? '',
        refererDomain: "http://mylekha.app",
      ));
    });

    test("create a transaction then check status pending", () async {
      final tranID = service.uniqueTranID();

      var _transaction = PaywayCreateTransaction(
        amount: 10.00,
        items: [
          PaywayTransactionItem(name: "ទំនិញ 1", price: 2, quantity: 1),
          PaywayTransactionItem(name: "ទំនិញ 2", price: 3, quantity: 1),
          PaywayTransactionItem(name: "ទំនិញ 3", price: 5, quantity: 1),
        ],
        reqTime: service.uniqueReqTime(),
        tranId: tranID,
        email: 'support@mylekha.app',
        firstname: 'Miss',
        lastname: 'My Lekha',
        phone: '010464144',
        option: PaywayPaymentOption.abapay_deeplink,
        shipping: 0.0,
        returnUrl: "https://mylekha.org/api/v1.0/integrate/payway/success",
        returnParams: EncoderService.base64_encode({ 'key_1': 'value_1', 'key_2': 'value_2' }),
        customFields :EncoderService.base64_encode({"Purcahse order ref":"Po-MX9901", "Customfield2":"value for custom field"}),
      );

      var createResponse =
          await service.createTransaction(transaction: _transaction, enabledLogger: true);

      expect(createResponse.abapayDeeplink != null, true,
          reason: "the deeplink should be a string according to docs");

      var checkResponse = await service.checkTransaction(
          transaction: PaywayCheckTransaction(
        tranId: _transaction.tranId,
        reqTime: service.uniqueReqTime(),
      ));

      expect(checkResponse.status == 2, true,
          reason: "the new transaction created should be pending or status 2");
    });

    test("generate checkout uri for a transaction then check status pending",
        () async {
      final tranID = service.uniqueTranID();

      var _transaction = PaywayCreateTransaction(
        amount: 10,
        items: [
          PaywayTransactionItem(name: "ទំនិញ 1", price: 5, quantity: 1),
          PaywayTransactionItem(name: "ទំនិញ 2", price: 3, quantity: 1),
          PaywayTransactionItem(name: "ទំនិញ 3", price: 2, quantity: 1),
        ],
        reqTime: service.uniqueReqTime(),
        tranId: tranID,
        email: 'support@mylekha.app',
        firstname: 'Miss',
        lastname: 'My Lekha',
        phone: '010464144',
        option: PaywayPaymentOption.abapay,
        shipping: 0.0,
        returnUrl: "https://mylekha.org/api/v1.0/integrate/payway/success",
        returnParams: EncoderService.base64_encode({ 'key_1': 'value_1', 'key_2': 'value_2' }),
        customFields :EncoderService.base64_encode({"Purcahse order ref":"Po-MX9901", "Customfield2":"value for custom field"}),
      );

      var webURI = await service.generateTransactionCheckoutURI(
          transaction: _transaction);
      io.File('checkout.html').writeAsString(webURI.data!.contentAsString());

      /// print(webURI);

      /// expect(
      ///     webURI.queryParameters['items'],
      ///     EncoderService.base64_encode(
      ///         _transaction.items.map((e) => e.toMap()).toList()),
      ///     reason: "encoded items should be equal");

      expect(webURI.toString().isNotEmpty, true,
          reason: "the uri should be generated");
    });
  });
}
